File: python/1026. Maximum Difference Between Node and Ancestor.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxAncestorDiff(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.res = -(1<<31)
        max_val = -(1<<31)
        min_val = 1<<31-1

        def dfs(root, max_val, min_val):
            if not root: return
            max_val = max(max_val, root.val)
            min_val = min(min_val, root.val)
            self.res = max(self.res, max(abs(max_val - root.val), abs(min_val - root.val)))
            dfs(root.left, max_val, min_val)
            dfs(root.right, max_val, min_val)
            
        dfs(root, max_val, min_val)
        return self.res

File: go/src/leetcode/1026/1026. Maximum Difference Between Node and Ancestor.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxAncestorDiff(root *TreeNode) (ans int) {
	var dfs func(node *TreeNode, mi, max int)
	dfs = func(node *TreeNode, mi, mx int) {
		if node == nil {
			return
		}
		ans = max(ans, max(node.Val-mi, mx-node.Val))
		mi, mx = min(mi, node.Val), max(mx, node.Val)
		if node.Left != nil {
			dfs(node.Left, mi, mx)
		}
		if node.Right != nil {
			dfs(node.Right, mi, mx)
		}
	}
	dfs(root, root.Val, root.Val)
	return
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

